### PR TITLE
Migrate material to primeng

### PR DESCRIPTION
**Pull request by belkassaby**

### CODE REVIEW

**Comment by belkassaby** [comment]
@ielis when you have time, let me know what you think for the metadata endpoint code

File: phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/PhenopacketLabMetadata.java [hunk 1]
```diff
@@ -1,13 +1,76 @@
...
+                this.uberonResource,
+                this.hgncResource,
+                this.ncitResource,
+                this.gssoResource));
+        return resourceList.stream();
+    }
+
+    public Optional<Resource> resourceByPrefix(String prefix) {
+        return switch (prefix.toUpperCase()) {
+            case "HP" -> Optional.of(hpoResource);
```
Comment (gerring) [inline phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/PhenopacketLabMetadata.java:63]: Could just use a map here rather than the switch but I don't think is matters.

File: phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/ValidateService.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package org.monarchinitiative.phenopacketlab.core;
+
+import org.phenopackets.phenopackettools.validator.core.ValidationResult;
+import org.phenopackets.phenopackettools.validator.core.ValidationResults;
+import org.phenopackets.phenopackettools.validator.core.ValidationWorkflowRunner;
+import org.phenopackets.phenopackettools.validator.jsonschema.JsonSchemaValidationWorkflowRunner;
+import org.phenopackets.schema.v2.PhenopacketOrBuilder;
+
+public class ValidateService {
```
Comment (gerring) [inline phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/ValidateService.java:9]: Should the Spring annotation be added to make this a service?

File: phenopacketlab-frontend/src/assets/img/sejal-desai.jpeg
Comment (gerring) [inline phenopacketlab-frontend/src/assets/img/sejal-desai.jpeg:1]: I assume sejal is okay with her image being added to the repository?

File: phenopacketlab-frontend/src/environments/environment.ts [hunk 3]
```diff
@@ -2,10 +2,12 @@
 // `ng build --prod` replaces `environment.ts` with `environment.prod.ts`.
 // The list of file replacements can be found in `angular.json`.
 
+const URL = 'http://localhost:8080';
 const API_URL = 'http://localhost:8080/api/v1';
```
Comment (gerring) [inline phenopacketlab-frontend/src/environments/environment.ts:6]: API_URL = URL+'/api/v1';

File: phenopacketlab-restapi/src/test/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                "      \"version\": \"2021-06-10\",\n" +
+                "      \"namespacePrefix\": \"NCBITaxon\",\n" +
+                "      \"iriPrefix\": \"http://purl.obolibrary.org/obo/NCBITaxon_\"\n" +
+                "    }],\n" +
+                "    \"phenopacketSchemaVersion\": \"2.0\"\n" +
+                "  }\n" +
+                "}";
+
+    private final String INCORRECT_PHENOPACKET =
+                "{\n" +
```
Comment (gerring) [inline phenopacketlab-restapi/src/test/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateControllerTest.java:221]: There could be a few more band cases in the tests. Are there any ways you can think of which could break the code with bad packets? Can tests pin these down?

**Review by gerring — APPROVED** [review]

File: phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/ValidateService.java (hunk 2 above)
Comment (belkassaby) [reply phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/ValidateService.java:9]: The bean is declared here  https://github.com/TheJacksonLaboratory/PhenopacketLab/pull/179/files/32ad8106a5bd244481c4bec7779824b38d3b61d6#diff-3336ffc5eae271b11cf6b9bbe07d71c7928465623113ac92820d53eca51bd0ac

File: phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/PhenopacketLabMetadata.java [hunk 5]
```diff
@@ -1,13 +1,76 @@
...
+            case "EFO" -> Optional.of(efoResource);
+            case "GENO" -> Optional.of(genoResource);
+            case "MONDO" -> Optional.of(mondoResource);
+            case "SO" -> Optional.of(soResource);
+            case "UBERON" -> Optional.of(uberonResource);
+            case "HGNC" -> Optional.of(hgncResource);
+            case "NCIT" -> Optional.of(ncitResource);
+            case "GSSO" -> Optional.of(gssoResource);
+            default -> Optional.empty();
+        };
```
[the remark this answers is not in the document]
Comment (belkassaby) [reply phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/PhenopacketLabMetadata.java:73]: Not sure we need hashcode and equals implementation here as even with different instances retrieved, the object returned from the endpoint will always be the same. @ielis what do you think?

File: phenopacketlab-frontend/src/assets/img/sejal-desai.jpeg
Comment (belkassaby) [reply phenopacketlab-frontend/src/assets/img/sejal-desai.jpeg:1]: Yes, she is (we forgot adding her to the list of contributors)

File: phenopacketlab-frontend/src/environments/environment.ts (hunk 3 above)
Comment (belkassaby) [reply phenopacketlab-frontend/src/environments/environment.ts:6]: good point

Comment (belkassaby) [reply phenopacketlab-frontend/src/environments/environment.ts]: Modified in [ef91e0](https://github.com/TheJacksonLaboratory/PhenopacketLab/pull/179/commits/ef91e0029799570f98014495acd15196609abe91)

File: phenopacketlab-restapi/src/test/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateControllerTest.java (hunk 4 above)
Comment (belkassaby) [reply phenopacketlab-restapi/src/test/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateControllerTest.java:221]: yeah there are a lot of possibilities as the schema is pretty large. I'll add a couple more smaller phenopackets testing different edge cases, but here Im not testing the validation (this is the job of phenopacket-tools on which I depend), Im just testing that the endpoint when it returns a non-validation.

File: phenopacketlab-io/src/main/java/org/monarchinitiative/phenopacketlab/io/PhenopacketResource.java [hunk 6]
```diff
@@ -4,11 +4,11 @@
 
 import java.util.Objects;
 
-class PhenopacketResource implements Resource {
+public class PhenopacketResource implements Resource {
```
Comment (ielis) [inline phenopacketlab-io/src/main/java/org/monarchinitiative/phenopacketlab/io/PhenopacketResource.java:7]: Why do we need this to be public?

Comment (ielis) [reply phenopacketlab-io/src/main/java/org/monarchinitiative/phenopacketlab/io/PhenopacketResource.java:7]: I think I see - for the Jackson serializer. This is a pain in the bum. However, I think I found a way how to fix this without making the implementation public. I'll commit the code. Please resolve this to acknowledge.. :)

File: phenopacketlab-restapi/src/main/java/org/monarchinitiative/phenopacketlab/restapi/controller/MetadataController.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    private final PhenopacketLabMetadata metadataService;
+
+    public MetadataController(PhenopacketLabMetadata metadataService) {
+        this.metadataService = metadataService;
+    }
+
+    @GetMapping("{prefix}")
+    public ResponseEntity<Resource> metadataByPrefix(@PathVariable("prefix") @Parameter(description = """
+                                                       __Accepted:__
```
Comment (ielis) [inline phenopacketlab-restapi/src/main/java/org/monarchinitiative/phenopacketlab/restapi/controller/MetadataController.java:28]: I think the accepted prefixes depend on what the `PhenopacketLabMetadata` offers. I would do this instead:
```java
    private final PhenopacketLabMetadata metadataService;
    private final List<String> acceptedPrefixes;

    public MetadataController(PhenopacketLabMetadata metadataService) {
        this.metadataService = metadataService;
        this.acceptedPrefixes = metadataService.resources()
                .map(Resource::getNamespacePrefix)
                .toList();
    }

...

@GetMapping("{prefix}")
    public ResponseEntity<Resource> metadataByPrefix(@PathVariable("prefix") @Parameter(description = """
WE NEED TO UPDATE THE DESCRIPTION..
""") String prefix) {
        if (!acceptedPrefixes.contains(prefix.toUpperCase())) {
            return ResponseEntity.badRequest().build();
        }
        return ResponseEntity.of(metadataService.resourceByPrefix(prefix));
    }
```

File: phenopacketlab-restapi/src/main/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateController.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+public class ValidateController {
+
+    private final ValidateService validateService;
+
+    public ValidateController(ValidateService validateService) {
+        this.validateService = validateService;
+    }
+
+    @PostMapping
+    public ResponseEntity<String> validatePhenopacket(@RequestBody String phenopacketString) {
```
Comment (ielis) [inline phenopacketlab-restapi/src/main/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateController.java:19]: I think we may need more detail here, not just one `String`. I think I'll need to do the same trick in Phenopacket tools with `jackson-annotations` in `provided` scope to implement automaitc JSON serialization of the validation results.

Comment (belkassaby) [reply phenopacketlab-restapi/src/main/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateController.java:19]: what type of extra detail do you think we would need? should the validate endpoint take more than just a phenopacket string?

Comment (ielis) [reply phenopacketlab-restapi/src/main/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateController.java:19]: Sorry, I meant output. The validation results can be more granular.

**Comment by belkassaby** [comment]
@ielis finally fixed the tests. Maybe not in the nicest way, but it works. what do you think?

File: phenopacketlab-restapi/src/main/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateController.java (hunk 8 above)
Comment (iimpulse) [reply phenopacketlab-restapi/src/main/java/org/monarchinitiative/phenopacketlab/restapi/controller/ValidateController.java:19]: We are also trying to make an API standard of things that should always be in a response. Could bring this up in that meeting.

File: phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/PhenopacketLabMetadata.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ielis) [reply phenopacketlab-core/src/main/java/org/monarchinitiative/phenopacketlab/core/PhenopacketLabMetadata.java]: I think @gerring was right at the time he was looking at this. Indeed there were many fields that were used neither in `hashCode` nor `equal`. However, we moved the fields elsewhere, leaving just one field `phenopacketSchemaVersion`. Thanks to this, the `hashCode` and `equals` should be OK. However, it would still be better to remove them. The class is supposed to be used as a singleton within Spring container and custom `equals` and `hashCode` look confusing in this context.

@belkassaby can you please remove `equals` and `hashCode` when you have a chance?

**Comment by belkassaby** [comment]
@ielis I will merge this PR into Development as it is becoming rather big. The PR is already deployed to SQA and DEV and the application can be viewed running at phenopacketlab-dev.jax.org and phenopacketlab-sqa.jax.org

**Comment by ielis** [comment]
Hi @belkassaby yes, please merge the PR.

**Review by ielis — APPROVED** [review]
Thank you, let's merge.
